### PR TITLE
fix(core): fix O3 table write, which can produce 0 values on column files over 2Gb

### PR DESCRIPTION
**Pull request by ideoma**
Fixes https://github.com/questdb/questdb/issues/2511

### CODE REVIEW

File: core/src/main/c/share/files.h [hunk 1]
```diff
@@ -294,6 +294,8 @@ JNIEXPORT jboolean JNICALL Java_io_questdb_std_Files_setLastModified
  * Signature: (JJ)I
  */
 
+#define MAX_RW_COUNT 0x7ffff000
```
Comment (puzpuzpuz) [inline core/src/main/c/share/files.h:297]: nit: would be nice to add the following text from `man 2 read` as a clarifying comment for this constant:
```
On  Linux,  read() (and similar system calls) will transfer at most 0x7ffff000 (2,147,479,552) bytes, returning the number of bytes actually transferred.
```

**Review by bluestreak01 — CHANGES_REQUESTED** [review]
We also need freebsd binary - happy to walk thru building process?

File: core/src/main/c/share/files.c [hunk 2]
```diff
@@ -50,7 +50,27 @@ JNIEXPORT jlong JNICALL Java_io_questdb_std_Files_write
...
-    return pwrite((int) fd, (void *) (address), (size_t) len, (off_t) offset);
+
+    off_t writeOffset = offset;
+    ssize_t written;
+
+    do {
+        size_t count = len > MAX_RW_COUNT ? MAX_RW_COUNT : len;
+        written = pwrite((int) fd, (void *) (address), count, writeOffset);
+        if (written < 0
+            // Signals should not interrupt sendfile on Linux but just to align with POSIX standards
```
Comment (puzpuzpuz) [inline core/src/main/c/share/files.c:61]:
```suggestion
            // Signals should not interrupt pwrite on Linux but just to align with POSIX standards
```

File: core/src/main/c/share/files.c [hunk 3]
```diff
@@ -85,7 +105,27 @@ JNIEXPORT jlong JNICALL Java_io_questdb_std_Files_read
...
 
-    return pread((int) fd, (void *) address, (size_t) len, (off_t) offset);
+    off_t readOffset = offset;
+    ssize_t read;
+
+    do {
+        size_t count = len > MAX_RW_COUNT ? MAX_RW_COUNT : len;
+        read = pread((int) fd, (void *) (address), count, readOffset);
+        if (read < 0
+            // Signals should not interrupt sendfile on Linux but just to align with POSIX standards
```
Comment (puzpuzpuz) [inline core/src/main/c/share/files.c:115]:
```suggestion
            // Signals should not interrupt pread on Linux but just to align with POSIX standards
```

File: core/src/main/c/share/files.h (hunk 1 above)
Comment (ideoma) [reply core/src/main/c/share/files.h:299]: done

File: core/src/main/c/share/files.c [hunk 4]
```diff
@@ -50,7 +50,20 @@ JNIEXPORT jlong JNICALL Java_io_questdb_std_Files_write
...
          jlong len,
          jlong offset) {
-    return pwrite((int) fd, (void *) (address), (size_t) len, (off_t) offset);
+
+    off_t writeOffset = offset;
+    ssize_t written = 1;
+    while (len > 0 && written > 0) {
+        size_t count = len > MAX_RW_COUNT ? MAX_RW_COUNT : len;
+        written = pwrite((int) fd, (void *) (address), count, writeOffset);
+        if (written < 0) {
```
[the remark this answers is not in the document]
Comment (ideoma) [reply core/src/main/c/share/files.c]: ok, done

**Review by puzpuzpuz — COMMENTED** [review]
Should be good to go once the BSD binary is updated.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:disappointed: **fail** : 2 / 5 (40.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/O3CopyJob.java|2|5|40.00%|

**Review by puzpuzpuz — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:disappointed: **fail** : 2 / 5 (40.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/O3CopyJob.java|2|5|40.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:disappointed: **fail** : 2 / 5 (40.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/O3CopyJob.java|2|5|40.00%|

**Review by bluestreak01 — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:disappointed: **fail** : 2 / 5 (40.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/O3CopyJob.java|2|5|40.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 22 / 22 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/O3CopyJob.java|22|22|100.00%|

**Review by bluestreak01 — APPROVED** [review]

**Pull request merged**